rust/otel-arrow-rust/benches/item_count.rs

```diff
@@ -0,0 +1,104 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
+
+//! This crate benchmarks counting log records in an OTLP LogsData.
+
+use criterion::{Criterion, black_box, criterion_group, criterion_main};
+
+use otel_arrow_rust::pdata::otlp::ItemCounter;
+use otel_arrow_rust::pdata::otlp::LogsVisitor;
+
+use otel_arrow_rust::proto::opentelemetry::common::v1::*;
+use otel_arrow_rust::proto::opentelemetry::logs::v1::*;
+use otel_arrow_rust::proto::opentelemetry::resource::v1::*;
+
+fn create_logs_data() -> LogsData {
+    let mut rl: Vec<ResourceLogs> = vec![];
+
+    for _ in 0..10 {
+        let kvs = vec![
+            KeyValue::new("k1", AnyValue::new_string("v1")),
+            KeyValue::new("k2", AnyValue::new_string("v2")),
+        ];
+        let res = Resource::new(kvs.clone());
+        let mut sls: Vec<ScopeLogs> = vec![];
+        for _ in 0..10 {
+            let is1 = InstrumentationScope::new("library");
+
+            let mut lrs: Vec<LogRecord> = vec![];
+
+            for _ in 0..10 {
+                let lr = LogRecord::build(2_000_000_000u64, SeverityNumber::Info, "event1")
+                    .attributes(kvs.clone())
+                    .finish();
+                lrs.push(lr);
+            }
+
+            let sl = ScopeLogs::build(is1.clone())
+                .log_records(lrs.clone())
+                .schema_url("http://schema.opentelemetry.io")
+                .finish();
+            sls.push(sl);
+        }
+        rl.push(ResourceLogs::build(res).scope_logs(sls).finish())
+    }
+
+    LogsData::new(rl)
+}
+
+fn count_logs(c: &mut Criterion) {
+    let mut group = c.benchmark_group("OTLP Logs counting");
+
+    let logs = create_logs_data();
+    assert_eq!(
+        1000,
+        ItemCounter::new().visit_logs(&LogsDataMessageAdapter::new(&logs))
+    );
+
+    _ = group.bench_function("Visitor", |b| {
+        b.iter(|| {
+            _ = black_box(ItemCounter::new().visit_logs(&LogsDataMessageAdapter::new(&logs)));
+        })
+    });
+
+    _ = group.bench_function("Manual", |b| {
+        b.iter(|| {
+            let mut count = 0;
+            for rl in &logs.resource_logs {
+                for sl in &rl.scope_logs {
+                    // Note! This is an optimization not available to the visitor.
+                    count += sl.log_records.len();
+                }
+            }
+            _ = black_box(count);
+        })
+    });
+
+    _ = group.bench_function("FlatMap", |b| {
+        b.iter(|| {
+            let count = logs
+                .resource_logs
+                .iter()
+                .flat_map(|rl| &rl.scope_logs)
+                .flat_map(|sl| &sl.log_records)
```

Comment:
Similar to the for-loop approach, we could simply sum up the `log_records.len()` instead counting each `log_record`. The compiler may already be doing that optimization, but I think we should still check in the optimized code.

```rust
let count = logs
    .resource_logs
    .iter()
    .flat_map(|rl| &rl.scope_logs)
    .map(|sl| sl.log_records.len())
    .sum::<usize>();
```